docker/ccf-pdo.sgx.yaml

```diff
@@ -0,0 +1,51 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+# This docker-compose file extends the basic sawtooth-pdo template with support
+# for SGX in hardware mode. To use add a '-f sawtooth-pdo.sgx.yaml' _after_ the
+# '-f sawtooth-pdo.yaml'. This can also be combined with sawtooth-pdo.local-code.yaml.
+
+# Before you can run the containers in hardware mode, you will have to prepare following
+# files in the sgx subdirectory
+# - sgx_spid.txt
+# - sgx_spid_api_key.txt
+# - sgx_ias_key.pem
+# See 'build/common-config -h' for information on the content of these files
+#
+# in the pdo-build shell (see sawtooth-pdo.yaml), before doing any operation
+# you also will have to register sgx-related information in the ledger with
+# following steps
+#    export PDO_SGX_KEY_ROOT=/project/pdo/build/opt/pdo/etc/keys/sgx/
+#    unset PDO_SPID PDO_SPID_API_KEY
+#    source /project/pdo/src/private-data-objects/build/common-config.sh
+#    make -C /project/pdo/src/private-data-objects/build conf register
+
+version: "2.1"
+
+services:
+
+  # PDO EHS, PS and client ...
+  pdo-build:
+    image: pdo-sgx-build
+    container_name: ccf-pdo-sgx-build
+    build:
+      args:
+        SGX_MODE: HW
+    volumes:
+      - ${PDO_SGX_KEY_ROOT:-./sgx/}:/project/pdo/build/opt/pdo/etc/keys/sgx/
+      - /var/run/aesmd:/var/run/aesmd
+    devices:
```

Comment:
i get that this "works. I do not understand why its necessary. what is so fundamentally different with this build over the sim build that it requires a completely new docker file?

Reply:
thank you, this file is the equivalent of sawthooth-pdo.sgx.yaml. And by adding this file, we are being consistent with what is done to enable end-to-end docker testing in HW mode for Sawtooth. Basically, this is treated as docker-compose options-file (specified as input via -f). This does two things 1. mounts the SGX keys, and SGX drivers on a sgx enabled machines, and overrides the image/container name (with pdo-sgx-build) during build process. I was also wondering why it is done like this for quite some time, but the more i read docker-compose documentation, what is originally done for Sawtooth is indeed the official recommended way to create optional parameters for marginally different builds. So, Just following what is originally there for Sawtooth.

Reply:
I agree with both of you, namely:
* it's (kind of) consistent with what we have with sawtooth
* we should double-check whether another docker image (built by docker compose, and thus only appearing in the CI, and not in the actual build) is necessary when we build in HW mode -- I don't think we can remove the docker compose file since we have to mount the necessary volumes.

One more thing regarding the container name: is there anything specific to ccf here?
In #379, I had called this file `pdo.sgx.yaml` specifically because it's about SGX HW mode, but it's ledger agnostic.

Reply:
@bvavala Thank you for the note. On your question as to why "create a separate file for ccf", IMO, this is the docker-compose options file that you would extend if you want to run the TP itself in HW mode. So, once we do that, the files will look meaningfully different for CCF and Sawtooth. Also, please note the container names are different. So, allows you to have both containers deployed at the same time, and just set PDO_LEDGER_URL to the TP that you use. Not possible if you collapse it to one file.

Reply:
"because we do it for sawtooth" is hardly justification. or more accurately... we should NEVER use "sawtooth does it" as justification for why we are making choices. 

we have four configurations for ccf that have to be addressed at some level:
1) CCF ledger is running in SIM mode, ESERVICE is running in SIM mode (what we mostly do now)
2) CCF ledger is running in SIM mode, ESERVICE is running in HW mode (more or less what we are adding)
3) CCF ledger is running in HW mode, ESERVICE is running in SIM mode (not obviously useful)
4) CCF ledger is running in HW mode, ESERVICE is running in HW mode (production)

Note that only the first two configurations matter for sawtooth. 

The compilation of CCF & ESERVICE should be exclusively based on CMakeFIle variables. NOT the docker images.
The docker images set up the necessary operating environment.